### PR TITLE
Replace tfgen result summaries

### PR DESCRIPTION
**Pull request by VenelinMartinov**
Should address #1348.  
Replaces the current summaries:  
```
General metrics:
	1241 total resources containing 13068 total inputs.
	511 total functions.
	0 entities are missing docs entirely because they could not be found in the upstream provider.

Description metrics:
	0 entity descriptions contained an <elided> reference and were dropped, including examples.
	0 entity descriptions contained an <elided> reference and were dropped, but examples were preserved.

Example conversion metrics:
	95 HCL examples failed to convert in all languages
	59 HCL examples were converted in at least one language but failed to convert to TypeScript
	59 HCL examples were converted in at least one language but failed to convert to Python
	59 HCL examples were converted in at least one language but failed to convert to Go
	59 HCL examples were converted in at least one language but failed to convert to C#
	73 entity document sections contained unexpected HCL code snippets. Examples will be converted, but may not display correctly in the registry, e.g. lacking tabs.

Argument metrics:
	9594 argument descriptions were parsed from the upstream docs
	196 top-level input property descriptions came from an upstream attribute (as opposed to an argument). Nested arguments are not included in this count.
	0 arguments contained an <elided> reference and had their descriptions dropped.
	0 nested arguments contained an <elided> reference and had their descriptions dropped.
	542 of 13068 resource inputs (4.15%) are missing descriptions in the schema

Attribute metrics:
	0 attributes contained an <elided> reference and had their descriptions dropped.
```

with the summaries from `examples_coverage_tracker`:

```
Additional example conversion stats are available by setting COVERAGE_OUTPUT_DIR.
Provider:     azuread
Success rate: 97.75% (522/534)

Converted 97.75% of csharp examples (87/89)
Converted 97.75% of go examples (87/89)
Converted 98.88% of java examples (88/89)
Converted 97.75% of python examples (87/89)
Converted 97.75% of typescript examples (87/89)
Converted 96.63% of yaml examples (86/89)
```

I've also added a message about the detailed summaries since it might be useful for people to find out about the additional stats available:  
`Additional example conversion stats are available by setting COVERAGE_OUTPUT_DIR.`.  

Happy to remove it, let me know if I should.

### CODE REVIEW

**Comment by t0yv0** [comment]
This is awesome. Consider hybridizing the display to still show some of the old numbers that are not present in the new display that seem a little useful, I'm thinking these things:

```
General metrics:
	1241 total resources containing 13068 total inputs.
	511 total functions.

Argument metrics:
	9594 argument descriptions were parsed from the upstream docs
	196 top-level input property descriptions came from an upstream attribute (as opposed to an argument). Nested arguments are not included in this count.
	0 arguments contained an <elided> reference and had their descriptions dropped.
	0 nested arguments contained an <elided> reference and had their descriptions dropped.
	542 of 13068 resource inputs (4.15%) are missing descriptions in the schema
```

We can probably drop "<elided>" lines if N=0.

**Comment by t0yv0** [comment]
@jkodroff WDYT? Does this look better? CC @ringods

**Review by t0yv0 — APPROVED** [review]

File: pkg/tfgen/examples_coverage_exporter.go [hunk 1]
```diff
@@ -440,22 +440,20 @@ func (ce *coverageExportUtil) exportMarkdown(outputDirectory string, fileName st
...
-// The fourth mode, which simply gives the provider name, and success percentage.
-func (ce *coverageExportUtil) exportHumanReadable(outputDirectory string, fileName string) error {
-
-	// The Coverage Tracker data structure is flattened to gather statistics about each language
-	type LanguageStatistic struct {
-		Total     int
-		Successes int
-	}
+// The Coverage Tracker data structure is flattened to gather statistics about each language
+type LanguageStatistic struct {
```
Comment (t0yv0) [inline pkg/tfgen/examples_coverage_exporter.go:444]: Beware when you lift structs to top-level with names starting from Capital Letter they get exported/public.

If you do not need it public call it "languageStatistic".

**Comment by jkodroff** [comment]
The new language conversion metrics look really nice! A few notes:

From a diagnostic perspective, IMO it's important to note when an example fails to convert in all languages versus in a particular language because the cause is usually different: Failing to convert in _all_ languages is usually because the source HCL is bad for one reason or another (invalid, references a provider we don't have, etc) and it relatively common. If it fails in _one_ language, it's often a codegen issue of some sort (or at least it's not b/c the source HCL was invalid) and can be an indication of a deeper issue.

Agree with @t0yv0 that the other metrics are useful, which is a predictable opinion coming from me since I added them in the first place)! Some reasons why we track this stuff (which should be in comments in the code so that they are more obvious in the future):

- Missing arguments matter to end users in the docs and is necessary if we change the argument parsing behavior to know whether we made things better or worse on the whole. (Docs parsing is basically impossible to get perfect and we often have to make changes that might make things better overall but worse in some circumstances.)
- The "top-level input property descriptions came from" is tracked because it's weird behavior that I am not sure is necessarily correct - they may be an indication that the TF schema is actually wrongly defined. (I don't remember the details on this, so don't take my word for it. IIRC it's most common in the AWS provider.)
- I don't know why I didn't add tracking for the number of inputs to functions. Seems like it would make sense to add, but IDK.
- Examples that lack tabs (assuming they still render wrong in the registry) is an important quality metric, so I recommend we keep it. It looks bad to the end-user, so if we can get that number down (zero would be awesome), we've significantly improved the end-user experience.
- Short-term, yes - keep the reference to `COVERAGE_OUTPUT_DIR`. Long-term those two things should be unified. Ideally, I wanted event-based logging with aggregator functions so that we could source all the information tracked in the `COVERAGE_OUTPUT` stuff, log to stdout, and aggregate stats from the same place. No suitable logging library existed when I checked over a year ago. The newer Go logging stuff may do what would be needed, so could be worth another look.

**Comment by t0yv0** [comment]
Thank for your input @jkodroff ! @VenelinMartinov let's review together again on Monday to make sure we're not removing any metrics Josh finds useful, then we can merge. The other suggestions here that are currently missing we can file for follow up. Thanks!

**Comment by ringods** [comment]
From a CI perspective, a shorter summary which is part of the build logs is handy as a quick check. I leave the implementation aspect of more detailed report to you (e.g. `COVERAGE_OUTPUT_DIR` or otherwise). As a maintainer of a number of bridged providers, I want the ability to generate a file with all the information that I can save as a workflow artifact. I can download that easily for further investigation.

GH docs:  https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts

**Pull request merged**